### PR TITLE
TECH-1634: Add new entry point for publication job subscription and deprecate old one.

### PR DESCRIPTION
**Pull request by cedmail**
## JIRA

https://jira.jahia.org/browse/TECH-1634

### CODE REVIEW

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/publication/PublicationJobSubscriptionExtension.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        if (bundleContext == null && eventHandlerServiceRegistrationDone == null) {
+            bundleContext = ((DXGraphQLProvider) BundleUtils.getOsgiService(GraphQLProvider.class, null)).getBundleContext();
+            new PublicationJobSubscriptionExtension().registerEventHandler();
+        }
+        return Flowable.create(obs -> {
+            String name = UUID.randomUUID().toString();
+            listeners.put(name, obs);
+            logger.debug("Registered publication job listener {}", name);
+            obs.setCancellable(() -> {
+                logger.debug("Registered publication job listener {}", name);
```
Comment (jkevan) [inline graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/publication/PublicationJobSubscriptionExtension.java:48]: Registered to be changed for Unregistered

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/publication/PublicationJobSubscriptionExtension.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        props.put(EventConstants.EVENT_TOPIC, Constants.CLUSTER_BROADCAST_TOPIC_PREFIX + "/publication/done");
+
+        eventHandlerServiceRegistrationDone = bundleContext.registerService(EventHandler.class, this, new MapToDictionary(props));
+        props.put(EventConstants.EVENT_TOPIC, Constants.CLUSTER_BROADCAST_TOPIC_PREFIX + "/publication/start");
+        props.put(org.osgi.framework.Constants.SERVICE_PID, PublicationJobSubscriptionExtension.class.getName() + ".EventHandler.Start");
+
+        eventHandlerServiceRegistrationStart = bundleContext.registerService(EventHandler.class, this, new MapToDictionary(props));
+    }
+
+    private void unregisterEventHandler() {
```
[the remark this answers is not in the document]
Comment (cedmail) [reply graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/publication/PublicationJobSubscriptionExtension.java:81]: it's not called yet as we do not have a solution to plug on graphql entrypoint lifecycle

**Review by jkevan — CHANGES_REQUESTED** [review]
It seem's we can listen on multiple topics for a single EventHandler, this could simplfy your implem by using a single registration.

Also I would suggest to separate the Subscription class from the OSGI EventHandler. It would improve readability of the implem and also allow to handle service registration more naturally:

```
@Component(property= { EventConstants.EVENT_TOPIC + "=" + Constants.CLUSTER_BROADCAST_TOPIC_PREFIX + "/publication/done",
                       EventConstants.EVENT_TOPIC + "=" + Constants.CLUSTER_BROADCAST_TOPIC_PREFIX + "/publication/start"})
public class PublicationJobEventListener implements EventHandler {

  @Override
  public void handleEvent(final Event event) {
        PublicationJobSubscriptionExtension.notifyListeners(event);    
  }
}
```

then in the PublicationJobSubscriptionExtension:
```
protected static void notifyListeners(Event event) {
        GqlPublicationEvent gqlPublicationEvent = new GqlPublicationEvent(
                event.getTopic().endsWith("/publication/done") ? GqlPublicationEvent.State.FINISHED : GqlPublicationEvent.State.STARTED,
                (List<String>) event.getProperty("siteKey"),
                (List<String>) event.getProperty("language"),
                (List<String>) event.getProperty("paths"),
                (List<String>) event.getProperty("user"));
        listeners.forEach((name, emitter) -> emitter.onNext(gqlPublicationEvent));
    }
```

This way the OSGI registration would be handle automatically.

**Review by jkevan — CHANGES_REQUESTED** [review]

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/DXGraphQLProvider.java [hunk 3]
```diff
@@ -403,4 +403,8 @@ private void transformCodeRegistryWithSDLDefaultFetcherFactory() {
         }).build();
         sdlSchemaService.setCodeRegistry(codeRegistry);
     }
+
+    public BundleContext getBundleContext() {
+        return bundleContext;
+    }
```
Comment (jkevan) [inline graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/DXGraphQLProvider.java:409]: Can be removed now.

**Review by jkevan — CHANGES_REQUESTED** [review]

**Review by jkevan — APPROVED** [review]
Thx for the modifications !

**Pull request merged**